File: microservices/l5-business-logic/product-svc/src/main/java/com/ilbuy/product/model/entity/Product.java
package com.ilbuy.product.model.entity;

import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.Instant;

@Entity
@Table(
    name = "product",
    indexes = {
        @Index(name = "idx_product_canonical_id", columnList = "canonical_id", unique = true),
        @Index(name = "idx_product_category_id",  columnList = "category_id"),
        @Index(name = "idx_product_name",          columnList = "name"),
        @Index(name = "idx_product_enabled",       columnList = "enabled")
    }
)
@EntityListeners(AuditingEntityListener.class)
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Canonical / universal product identifier shared across platforms.
     * e.g. "PROD-iphone15pro-256gb-black"
     */
    @Column(name = "canonical_id", nullable = false, unique = true, length = 128)
    private String canonicalId;

    @Column(nullable = false, length = 256)
    private String name;

    @Column(length = 4096)
    private String description;

    @Column(name = "brand", length = 128)
    private String brand;

    @Column(name = "category_id")
    private Long categoryId;

    /** Representative image URL */
    @Column(name = "image_url", length = 512)
    private String imageUrl;

    /** Reference / MSRP price for display */
    @Column(name = "reference_price", precision = 14, scale = 2)
    private BigDecimal referencePrice;

    @Column(nullable = false)
    @Builder.Default
    private Boolean enabled = true;

    /** Comma-separated tags for search */
    @Column(name = "tags", length = 1024)
    private String tags;

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private Instant updatedAt;
}

File: mall/models/product.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional


@dataclass
class Product:
    """商品模型"""
    name: str
    price: float
    stock: int
    category_id: int
    description: str = ''
    image_url: str = ''
    id: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    def to_dict(self) -> dict:
        return {
            'id': self.id,
            'name': self.name,
            'price': self.price,
            'stock': self.stock,
            'category_id': self.category_id,
            'description': self.description,
            'image_url': self.image_url,
            'created_at': str(self.created_at) if self.created_at else None,
            'updated_at': str(self.updated_at) if self.updated_at else None,
        }

    @staticmethod
    def from_row(row: tuple) -> 'Product':
        product = Product(
            id=row[0],
            name=row[1],
            price=row[2],
            stock=row[3],
            category_id=row[4],
            description=row[5] or '',
            image_url=row[6] or '',
        )
        product.created_at = row[7]
        product.updated_at = row[8]
        return product
